ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+
+This abstraction also enable Konflux to implement other features in the future like for example limmiting the amount of workspaces per user, allowing workspace delete self-service and adding workspace delete protection.
+
+
+### Upstream
+
+There is not support for multi-cluster for a single instance of Konflux. The Idea here is to require a new deployment of konflux of there is a need to scale beyond a single cluster. Therefore, there is a one to one relationship between konflux instance and k8s cluster.
+
+With this contraint in place workspace concept can be implemented differently
+We relay on the ability to create namespaces directly. Which requires access to all authenticated users to have premissions to create a namespace.
+
+The boostrapping of the namespace created is delegate to a GitOps workflow that will automatically create an application with the necesary resources, including RoleBinding and Quotas and will continue to synch this through the lifecycle of the project.
+
+To be able to see exisiting workspaces for a specific user, we developed a small abstraction called workspace manager, which is able to determine what namepsace have a rolebinding providing access to the current user.
+
+There is currently no way to delete a workspace self-service. User will have to make a request to a konflux admin to delete a specific workspace
+
+### Namespace Limitations
+
+Kubernetes namespaces where not designed for self-service provisioning in a multi-tenant.
+
+Access for namespace list and create operation is all namepaces or none. This affects multi-tenancy if self-service is desired.
+
+Access to get and delete can be granted via RBAC once the namespace is created. This will required admin access which also affect self-service.
+
+Allowing access to specific Roles or ClusterRoles via RoleBinding can be done by an Admin, also braking self-service. This is required to be able to execute Konflux actions inside the namespace.
+
+## Decision
+
+Introduce a workspace CRD and corresponding controller as a project of our community.
+
+We will provide all authenticated users access to create their own workspaces within a limit.
+They will also be able to update, list, get and delete only workspaces they own. 
+
+## Consequences
+
+Extending k8s API with the workspace concept will allow konflux user to manage their konflux workspaces in a secured, self-service manner while konflux admin can maintain multi-tenancy.
+
+Such extention of the API that wraps namepsace has precedence, for example OpenShift apis provide a Project kind the wraps a k8s namespace.
+
+### Additional benefit 
+
+We can automate the bootstraping of namespaces with extra resources a konflux admin deems proper for their specific use case.
+
+We can create exension points that enable the community to use different backends of the implementation which will include the ability to scale to multiple cluster with a single instance of Konflux.
```

Comment:
1. Do we have any requirement today for multi-cluster extensions or we can wait until we will have evidence that it's needed?
2. When you are saying "we", to who you refer? If I'm a contributor of Konflux, how do I develop an extension? to where I submit the code and who needs to review it? how the workspace abstraction helps me to develop an extension?

Reply:
> 1. Do we have any requirement today for multi-cluster extensions or we can wait until we will have evidence that it's needed?
We do not have to create a backend that support multi-cluster. But the ability to extend the controler with your own backend is what the value is here. You have the ability to keep your user API and scale more as needed and btw contribute your backend back to the community. 
> 2. When you are saying "we", to who you refer? If I'm a contributor of Konflux, how do I develop an extension? to where I submit the code and who needs to review it? how the workspace abstraction helps me to develop an extension?
The audience here is the Konflux community, so "we" is the audience for this ADR